### PR TITLE
Be able to use string type index in array, fix #5884

### LINKED ISSUE #5884: Vue.set Api strange behavior if path is a numerical string
### Version
2.3.4

### Reproduction link
[https://jsfiddle.net/v0dbnvv0/1/](https://jsfiddle.net/v0dbnvv0/1/)

### Steps to reproduce
Vue.set(array, "indexInTheArrayInString", value) is not working as expected compare to Vue.set(array, "indexInTheArrayInNumericForm", value)

### What is expected?
Both version working the same i suppose?

### What is actually happening?
Watcher on the value are not triggered even if the value is correctly changed in the model

---
My currently fixed in my code:
export function AdvVueSet(obj, path, value) {
    if (!isNaN(path)) {
        path =  +path;
    }
    Vue.set(obj, path, value);
}

### PR DESCRIPTION
**Pull request by Peter-WF**
https://github.com/vuejs/vue/issues/5884

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

File: src/core/observer/index.js [hunk 1]
```diff
@@ -189,7 +189,7 @@ export function defineReactive (
  * already exist.
  */
 export function set (target: Array<any> | Object, key: any, val: any): any {
-  if (Array.isArray(target) && typeof key === 'number') {
+  if (Array.isArray(target) && !isNaN(parseFloat(key))) {
```
Comment (posva) [inline src/core/observer/index.js]: I think something like a double `typeof...` (+ `key !== ''`) or `key !== '' && typeof Number(key) === 'number'` would be better. `parseFloat` looks like an overkill and `isNaN` is not supported on IE. wdyt @defcc ?

Comment (yyx990803) [inline src/core/observer/index.js]: `isNaN` is supported in IE. The problem is that `parseFloat` would turn strings like `"123ABC"` into `123`, which should probably be treated as string keys instead.

I think we should only support strings that are integers, something like `typeof key === 'number' || /^\d+$/.test(key)`

Comment (posva) [inline src/core/observer/index.js]: I see, `isNaN` exists as a global function but `Number.isNaN` does not exist 😨

Comment (Peter-WF) [inline src/core/observer/index.js]: Thank you all! 
It'd be a good idea to replace `parseFloat` with `regular`

**Pull request merged**